File: LinkedList2.js
class Node{
    constructor(value){
        this.value=value;
        this.next=null;
    }
}

class LinkedList{
    constructor(){
        this.head=null,
        this.tail=null,
        this.idx=0
    }

    append(value) {
        const apnode=new Node(value);
        if(this.idx===0){
            this.head=apnode;
            this.tail=apnode;
        }
        else if ( this.idx===1){
            this.head.next=apnode;
            this.tail=apnode;
        }
        else{           
            this.tail.next=apnode;
            this.tail=apnode;
        }
        this.idx++;
    }

    printlist(){
        let tnode=this.head;
        for ( var i=0;i<this.idx;i++){
            console.log(tnode.value);
            tnode=tnode.next;
        }
    }

    search(value){
        let tmpnode=this.head;
        for (var i=0 ; i < this.idx ; i++){
            if(tmpnode.value==value){
                return i+1;
            }
            else{
                tmpnode=tmpnode.next;
            }
        }
        return false;
    }

    delete(idx){
        let bnode=this.head;
        let tnode=this.head;
        if(idx===1){
            this.head=this.head.next;
        }
        else{
            for(var i=1;i<idx;i++){
                bnode=tnode;
                tnode=tnode.next;
            }
            bnode.next=tnode.next;
        }
        if(idx===this.idx){
            this.tail=bnode;
        }
        this.idx--;
        if (this.idx==0){
            this.head=null;
            this.tail=null;
        }

    }

    insert(idx,value){
        const insnode=new Node(value);
        let tmpnode=this.head;

        if (idx==1){
            insnode.next=this.head;
            this.head=insnode;
            if (this.idx==0){
                this.tail=insnode;
            }
        }
        else if(idx>this.idx){
            this.tail.next=insnode;
            this.tail=insnode;
        }
        else{
            for( var i=2; i<this.idx;i++){
                tmpnode=tmpnode.next;
            }
            insnode.next=tmpnode.next;
            tmpnode.next=insnode;           
        }
        this.idx++;
    }

    select(idx){
        let tmpnode=this.head;
        for( var i=1;i<idx;i++){
            tmpnode=tmpnode.next;            
        }
        return tmpnode.value;
    }

    update(idx,value){
        let tmpnode=this.head;
        for(var i=1;i<idx;i++){
            tmpnode=tmpnode.next;
        }
        tmpnode.value=value;
    }



}
const l=new LinkedList();
l.append("kim");
l.append("hyung");
l.append("keun");
l.append("!!!");
//l.delete(4);
//l.delete(3);
//l.delete(2);
//console.log(l);
l.insert(5,"aa");
l.update(5,"aaa");
l.printlist();
console.log(l);
console.log(l.select(5));
console.log(l.search("!!"));

File: LinkedList2.pl
package Node;

use strict;
use warnings;

sub new
{
    my $class = shift;
    my $self = {};
    $self->{value} = shift;
    $self->{next}  = undef;
    bless $self , $class;
    return $self ;
}

package LinkedList;

sub new{

    my $class = shift;
    my $self = {};
    $self->{head} = undef;
    $self->{tail} = undef;
    $self->{idx} = 0;
    bless $self , $class;
    return $self ;
}

sub append {
    my ( $self, $value ) = @_ ;
    my $apnode = Node->new($value);
 
    if ( $self->{idx} == 0 ) {
        $self->{head} = $apnode;
        $self->{tail} = $apnode;
    }
    elsif ( $self->{idx} == 1 ){
        $self->{head}->{next} = $apnode;
        $self->{tail} = $apnode;
    }
    else{
        $self->{tail}->{next} = $apnode;
        $self->{tail} = $apnode;
    }
    $self->{idx} += 1;

}

sub printlist{
    my $self = shift;
    my $tnode = $self->{head};
    my $i=0;
    while ($i< $self->{idx}){
        print ($tnode->{value},"\n");
        $tnode = $tnode->{next};
        $i++;
    }
}

sub search{
    my ($self, $value) = @_;
    my $tnode = $self->{head};
    for (my $ii=1; $ii<=$self->{idx} ;$ii++){
        if ($tnode->{value} eq $value ){
            return $ii;
        }
        $tnode = $tnode->{next};
    }
    return -1
}

sub delete{
    my ($self, $idx) = @_;
    my $tnode = $self->{head};
    my $bnode = $self->{head};

    if ($idx == 1) {
        $self->{head} =$tnode->{next};
    }
    else {
        for (my $i=1 ; $i <$idx ; $i++){
            $bnode=$tnode;
            $tnode=$tnode->{next};
        }
        $bnode->{next}=$tnode->{next};
    }
    
    if($idx == $self->{idx}){
        $self->{tail}=$bnode;
    }
    
    $self->{idx} -=1;
    if ($self->{idx}==0){
        $self->{head} = undef;
        $self->{tail} = undef;
    }
}

sub insert{
    my ($self,$idx,$value) = @_;
    my $insnode = Node->new($value);
    my $tmpnode = $self->{head};
    if ($idx==1){
        $insnode->{next}=$self->{head};
        $self->{head}=$insnode;
        if( $self->{idx} == 0){
            $self->{tail}=$insnode;
        }
    }
    elsif($idx > $self->{idx}){
        $self->{tail}->{next}=$insnode
    }
    else {
        for( my $i=1 ; $i < $idx; $i++){
            $tmpnode = $tmpnode->{next};
        }
        $insnode->{next} = $tmpnode->{next};
        $tmpnode->{next}=$insnode;
    }
    $self->{idx} +=1;
}

sub select {
    my($self,$idx) = @_;
    my $tmpnode= $self->{head};
    for ( my $i=1; $i < $idx; $i++){
        $tmpnode=$tmpnode->{next};
    }
    return $tmpnode->{value}
}

sub update{
    my ($self,$idx,$value) = @_;
    my $tmpnode=$self->{head};
    for ( my $i=1; $i<$idx; $i++){
        $tmpnode = $tmpnode->{next};
    }
    $tmpnode->{value} = $value;

}

my $jlist = new LinkedList();

=xx
$jlist->insert(1,"aa");
$jlist->printlist();
=cut

#=comments
$jlist->append("kim");
$jlist->append("hyung");
$jlist->append("keun");
$jlist->append("!!!");
#$jlist->delete(5);
#$jlist->delete(3);
#$jlist->delete(2);
#$jlist->delete(1);
$jlist->update(5,"zz");
print ("tail:",$jlist->{tail}->{value},"|head:",$jlist->{head}->{value},"\n");
$jlist->printlist();

my $aa = $jlist->select(3);
print $aa;
 $aa = $jlist->select(2);
print $aa;
 $aa = $jlist->select(1);
print $aa;
